### ISSUE TITLE
how to build a spark trainingImage

### ISSUE THREAD

**Issue by data-compute-geek** [awslabs/amazon-sagemaker-examples#275]
How can I build a trainingImage for a custom spark algorithm as the kmeans trainingImage in amazon-sagemaker-examples/sagemaker-spark/pyspark_mnist/pyspark_mnist_custom_estimator.ipynb ?

Github has examples for bring-your-own-algorithm in scikit and r. Are there examples for spark? Or how can I see the contents of the kmeans trainingImage so I can use it as a template to create my own image?

Thanks a lot.

**Comment by lynnlangit** [comment]
My team is working on a project in this area now.  We are still in code review, so the code isn't public yet.  Probably the simplest way to stay updated is to follow me in GitHub, since I am not an AWS employee, so I am not sure if the team will incorporate our code into this Repo or not.  Our examples are with Spark in Scala.   HTH

**Comment by data-compute-geek** [comment]
Hi Lynn, 

Thank you for the info. Just followed you. Please also keep me posted in your convenience.

I wish AWS can share the kmeans docker image or at least the basics of creating a spark container.

**Comment by djarpin** [comment]
Hi @data-compute-geek .

Just to confirm... Are you interested in training a Spark algorithm and hosting a Spark model in SageMaker?  Or just hosting the Spark model?  If it's the latter, the [workshop](https://github.com/nmukerje/sagemaker-spark-worshop) I shared in [this issue](https://github.com/awslabs/amazon-sagemaker-examples/issues/276) should hopefully help.

If it's the former, I'd be interested in understanding your use case in more detail.  Is there a reason you'd rather train your Spark algorithm in SageMaker's environment?  Is it because SageMaker manages spinning up and down your training cluster?  Something else?

SageMaker's built-in k-means algorithm doesn't use Spark for processing.  It's a custom implementation that uses slightly different math than Spark's k-means algorithm, and uses the Apache MXNet framework for numerical processing in order to take advantage of its parameter server for distributed computation and hardware acceleration for GPU processing.

@lynnlangit - We'd love community commits, so please feel free to submit a pull request when you've got something to share.  Most of our contributions up to this point have been internal though, so bear with us if there are a few kinks in the processing of your PR.

Thanks.

**Comment by data-compute-geek** [comment]
Hi @djarpin ,

Thank you for the quick response and the clarification.
I just want to use SageMaker to quickly deploy my spark models with minimal extra data processing before the endpoint is invoked. Before you shared the workshop, my understanding was that in order to deploy a model through SageMaker, the natural way is to train the model through SageMaker. I actually have already trained a spark model, and our engineers are in the process of productionizing it which seems to involve a significant amount of effort.  Going forward, we will have many spark, mxnet, and scikit-learn models, and it is of high priority for me to explore SageMaker for both training and deployment for these incoming models in order to iterate fast.  

The training data for my current model contains a good number of text fields in addition to numerical fields,  a number of pre-processing steps need to be done before I can train the final ML model, and in production, I would like to be able to invoke the endpoint with the raw input and let the endpoint take care of both the pre-processing steps and the final model inferences. 

So here is my updated understanding on developing models with SageMaker. To train and deploy  mxnet and scikit-learn models, SageMaker can be easily configured to serve the purpose. In particular, the pre-processing steps can be done through things like Lambda/python in production. For spark models involving multiple spark transformers/estimators, the approach in the workshop is more appropriate.  Am I right?

Again, thank you so much for your help.

**Comment by djarpin** [comment]
I would say that SageMaker is a set of components that can be used independently or together.  So, you can train and host on SageMaker, just train on SageMaker, or train outside of SageMaker but still host in SageMaker.  And since SageMaker uses docker containers, any algorithms you can wrap up into docker (whether that's scikit, MXNet, or Spark) you can use on SageMaker.

SageMaker has developed some of its own algorithms as well as pre-built TensorFlow, MXNet, and now Chainer deep learning framework containers.  We also have our example notebook showing how to bring a scikit algorithm to SageMaker.

The workshop is meant to show how you can take a model trained in Spark and deploy to SageMaker for hosting.  This includes setting up the Dockerfile and serving code.

Where to handle pre-processing will depend on your use case.  You can do the pre-processing in a hosting container, or you may choose to do the pre-processing outside of your hosting container (e.g. Lambda, EMR, etc.).  In general, I think the mental model we've been working off with Spark is that users will have a Spark cluster and will do the pre-processing there before submitting their endpoint request to get the prediction from their machine learning model.  This doesn't have to be the case, but deviating from it might require a bit more customization depending on your specific problem.  Again, SageMaker is flexible enough to handle just about whatever you put in your Docker container.

Thanks.

**Comment by data-compute-geek** [comment]
Hi @djarpin ,

I am trying to duplicate the workshop, but the endpoint fails to be created at the Step 3 when I run the following code segment.
------------------------------------------------------------------------
## Create Sagemaker Endpoint
response = sagemaker.create_endpoint(
    EndpointName=endpoint_name,
    EndpointConfigName=endpoint_config_name,
)

print (response)
-----------------------------------------------------------------------

When I click on the endpoint, the aws console shows the following message:
Failure reason
Please make sure all images included in the model for the production variant default-variant-name exist, and that the execution role used to create the model has permissions to access them

I double check everything, and nothing appears to be wrong.

Would you please give me a hand?
Thanks a lot.

**Comment by djarpin** [comment]
Hi @data-compute-geek - Apologies for the lack of response until now.  Hopefully you've self-resolved, and an AWS Support Case may actually be the best way to go since they'll be able to investigate your permissions and setup more thoroughly.  For future reference, potential things to check might be:

1. Ensuring your model artifact, ECR image, and SageMaker endpoint are all in the same region.
1. Ensuring you have the proper SageMaker [permissions](https://docs.aws.amazon.com/sagemaker/latest/dg/using-identity-based-policies.html) on your IAM role.

**Comment by djarpin** [comment]
Closing due to inactivity.  Feel free to re-open if necessary.  Thanks.
